### ISSUE TITLE
Alertmanager fails sometimes to send alerts to prom2teams

### ISSUE THREAD

**Issue by wadhahbenaissacdt** [idealista/prom2teams#289]
# Description

I have the prometheus stack along side prom2teams installed using helm charts (appVersion: 3.2.2) in an AKS.
At first it was working fine, after some time, I started getting alerts in my MS-Teams channel that "AlertmanagerClusterFailedToSendAlerts".
This alert is triggered usually every 10, 15, 20 minutes, and each time, after 5minutes, it gets resolved automatically.
The logs from prom2teams show:
```
│ 2022-04-19 14:07:18,695 - prom2teams_app - ERROR - An unhandled exception occurred. Error performing request to: {}.                                                                       │
│  Returned status code: {}.                                                                                                                                                                 
│  Returned data: {}                                                                                                                                                                         
│  Sent message: {}                                                                                                                                                                          
│ Traceback (most recent call last):                                                                                                                                                         
│   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 2292, in wsgi_app                                                                                                       │
│     response = self.full_dispatch_request()                                                                                                                                               
│   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1815, in full_dispatch_request                                                                                          │
│     rv = self.handle_user_exception(e)                                                                                                                                                     
│   File "/usr/local/lib/python3.8/site-packages/flask_restplus/api.py", line 583, in error_router                                                                                           │
│     return original_handler(e)                                                                                                                                                             
│   File "/usr/local/lib/python3.8/site-packages/flask_restplus/api.py", line 583, in error_router                                                                                           │
│     return original_handler(e)                                                                                                                                                             
│   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1718, in handle_user_exception                                                                                          │
│     reraise(exc_type, exc_value, tb)                                                                                                                                                       
│   File "/usr/local/lib/python3.8/site-packages/flask/_compat.py", line 35, in reraise                                                                                                      │
│     raise value                                                                                                                                                                            
│   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1813, in full_dispatch_request                                                                                          │
│     rv = self.dispatch_request()                                                                                                                                                           
│   File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1799, in dispatch_request                                                                                               │
│     return self.view_functions[rule.endpoint](**req.view_args)                                                                                                                             │
│   File "/usr/local/lib/python3.8/site-packages/flask_restplus/api.py", line 325, in wrapper                                                                                                │
│     resp = resource(*args, **kwargs)                                                                                                                                                       
│   File "/usr/local/lib/python3.8/site-packages/flask/views.py", line 88, in view                                                                                                           │
│     return self.dispatch_request(*args, **kwargs)                                                                                                                                          │
│   File "/usr/local/lib/python3.8/site-packages/flask_restplus/resource.py", line 44, in dispatch_request                                                                                   │
│     resp = meth(*args, **kwargs)                                                                                                                                                           
│   File "/usr/local/lib/python3.8/site-packages/prom2teams/app/versions/v1/namespace.py", line 28, in post                                                                                  │
│     self.sender.send_alerts(alerts, app.config['MICROSOFT_TEAMS']['Connector'])                                                                                                            │
│   File "/usr/local/lib/python3.8/site-packages/prom2teams/app/sender.py", line 26, in send_alerts                                                                                          │
│     self.teams_client.post(teams_webhook_url, team_alert)                                                                                                                                  │
│   File "/usr/local/lib/python3.8/site-packages/prom2teams/app/teams_client.py", line 41, in post                                                                                           │
│     simple_post(teams_webhook_url, message)                                                                                                                                                │
│   File "/usr/local/lib/python3.8/site-packages/prom2teams/app/teams_client.py", line 35, in simple_post                                                                                    │
│     self._do_post(teams_webhook_url, message)                                                                                                                                              │
│   File "/usr/local/lib/python3.8/site-packages/prom2teams/app/teams_client.py", line 54, in _do_post                                                                                       │
│     raise MicrosoftTeamsRequestException(                                                                                                                                                  │
│ prom2teams.app.exceptions.MicrosoftTeamsRequestException: Error performing request to: {}.                                                                                                 │
│  Returned status code: {}. 
│  Returned data: {}                                                                                                                                                                        
│  Sent message: {}                                                                                                                                                                         
│
│ 2022-04-19 14:07:18,697 - werkzeug - INFO - 10.244.0.79 - - [19/Apr/2022 14:07:18] "POST / HTTP/1.1" 400 -
```

The logs from alertmanager show:
```
│ level=error ts=2022-04-19T14:07:18.697Z caller=dispatch.go:354 component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="prom2teams/webhook[0]: notify retry canceled due to unrecoverable error after 1 attempts: unexpected status code 400: http://opsstack-prom2teams:8089"
```

**Comment by blalop** [comment]
Hi @wadhahbenaissacdt ,

we are suffering this issue too in our deployment, it seems that Microsoft Teams sometimes answers with 503 errors and that the error is not in our side.

Anyways, an issue was found in 3.2.x releases concerning this logging. It that was already addressed in https://github.com/idealista/prom2teams/pull/279. This fix was released in 3.3.0 and should help you diagnose this issue.

**Comment by BarnesJeff** [comment]
Hi @wadhahbenaissacdt @blalop,

We were also getting failed alerts when using prom2teams with [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) as well.
After using the 3.3.0 prom2teams release we traced the problem to double quotes in an alert description resulting in invalid json ie.
`Pod stable-client1/stable-client1 (debug) is in waiting state (reason: "CrashLoopBackOff").,`

Looking at https://github.com/idealista/prom2teams/blob/d3712cbb96f2e5a9847b1bdf0cf27fcdca26b61b/prom2teams/teams/alert_mapper.py there is a function which removes double quotes but is only called when group_alerts_by is not set in the prom2teams config.

To fix this we deployed custom teams.j2 that uses jinja2 tojson filter to escape characters. This should fix class of errors where alerts fail due to invalid JSON. I'm not familiar enough with the project to know if this would break things, but if not let me know and I can submit PR.

```
{%- set
  theme_colors = {
    'resolved' : '2DC72D',
    'critical' : '8C1A1A',
    'severe' : '8C1A1A',
    'warning' : 'FF9A0B',
    'unknown' : 'CCCCCC'
  }
-%}

{
    "@type": "MessageCard",
    "@context": "http://schema.org/extensions",
    "themeColor": "{% if status=='resolved' %} {{ theme_colors.resolved }} {% else %} {{ theme_colors[msg_text.severity] }} {% endif %}",
    "summary": {% if status=='resolved' %} {{ ("(Resolved) " + msg_text.summary) | tojson }}{% else %}{{ msg_text.summary | tojson }} {% endif %},
    "title": "Prometheus alert {% if status=='resolved' %}(Resolved) {% elif status=='unknown' %} (status unknown) {% endif %}",
    "sections": [{
        "activityTitle": {{ msg_text.summary | tojson }},
        "facts": [{% if msg_text.name %}{
            "name": "Alert",
            "value": {{ msg_text.name | tojson }}
        },{% endif %}{% if msg_text.instance %}{
            "name": "In host",
            "value": {{ msg_text.instance | tojson }}
        },{% endif %}{% if msg_text.severity %}{
            "name": "Severity",
            "value": {{ msg_text.severity | tojson }}
        },{% endif %}{% if msg_text.description %}{
            "name": "Description",
            "value": {{ msg_text.description | tojson }}
        },{% endif %}{
            "name": "Status",
            "value": {{ msg_text.status | tojson }}
        }{% if msg_text.extra_labels %}{% for key in msg_text.extra_labels %},{
            "name": {{ key | tojson }},
            "value": {{ msg_text.extra_labels[key] | tojson }}
        }{% endfor %}{% endif %}
        {% if msg_text.extra_annotations %}{% for key in msg_text.extra_annotations %},{
            "name": {{ key | tojson }},
            "value": {{ msg_text.extra_annotations[key] | tojson }}
        }{% endfor %}{% endif %}],
          "markdown": true
    }]
}
```
